### PR TITLE
added posts crud

### LINKED ISSUE #2: Post CRUD endpoints on backend
Endpoints:

GET /posts
- This should return all the posts and it should be pageable.
- The list returned needs to be sorted by `createdAt`
- For pagination it should accept a `page` parameter for the current page index (starting from 0) and a `page_size` parameter for the number of items on each page

GET /posts/{postId}

POST /posts
- Creates a new post
- The backend needs to check if the current user has permission to do it

DELETE /posts/{postId}
- Deletes an existing post
- The backend needs to check if the current user has permission to do it

PATCH /posts/{postId}
- Modifies an existing post
- The backend needs to check if the current user has permission to do it

### PR DESCRIPTION
**Pull request by FearsomeRover**
Closes #2

### CODE REVIEW

File: apps/backend/src/posts/posts.service.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    });
+  }
+
+  async update(id: number, updatePostDto: UpdatePostDto, user: User) {
+    const currentPost = await this.prisma.post.findUnique({
+      where: {
+        id,
+      },
+    });
+    if (currentPost.authorId !== user.authSchId && user.role !== 'BODY_ADMIN') {
```
Comment (Isti01) [inline apps/backend/src/posts/posts.service.ts:50]: Is it a good idea to only allow the original author to edit/delete posts?

File: apps/backend/src/posts/posts.service.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    });
+  }
+
+  async remove(id: number, user: User) {
+    const currentPost = await this.prisma.post.findUnique({
+      where: {
+        id,
+      },
+    });
+    if (currentPost.authorId !== user.authSchId && user.role !== 'BODY_ADMIN') {
```
Comment (Isti01) [inline apps/backend/src/posts/posts.service.ts:67]: Same here

File: apps/backend/src/posts/posts.service.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    });
+  }
+
+  async remove(id: number) {
+    return this.prisma.post.delete({
+      where: {
+        id,
+      },
+    });
+  }
```
Comment (Tschonti) [inline apps/backend/src/posts/posts.service.ts:59]: prisma update and delete methods throw an exception if no record is found, we should catch so no 500 error is thrown

File: apps/backend/src/posts/posts.controller.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  update(@Param('id') id: string, @Body() updatePostDto: UpdatePostDto) {
+    return this.postsService.update(Number(id), updatePostDto);
+  }
+
+  @Roles(Role.BODY_ADMIN, Role.BODY_MEMBER)
+  @ApiBearerAuth()
+  @Delete(':id')
+  remove(@Param('id') id: string) {
+    return this.postsService.remove(Number(id));
+  }
```
Comment (Tschonti) [inline apps/backend/src/posts/posts.controller.ts:47]: I think these also need to be annotated with the `@UseGuards(AuthGuard('jwt'))`

File: apps/backend/src/posts/posts.service.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        },
+      });
+    } catch (e) {
+      if (e instanceof Prisma.PrismaClientKnownRequestError) {
+        if (e.code === 'P2025') {
+          throw new NotFoundException('This post does not exist.');
+        }
+        throw new InternalServerErrorException('An error occurred.');
+      }
+    }
```
Comment (Tschonti) [inline apps/backend/src/posts/posts.service.ts:50]: classic mistake: trying to catch an error in an async call without awaiting it

File: apps/backend/src/posts/posts.service.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        switch (e.code) {
+          case 'P2000':
+            throw new BadRequestException('The content is too long.');
+          case 'P2025':
+            throw new NotFoundException('This post does not exist.');
+        }
+      }
+      throw new InternalServerErrorException('An error occurred.');
+    }
+  }
```
Comment (Tschonti) [inline apps/backend/src/posts/posts.service.ts:92]: same

**Review by Tschonti — CHANGES_REQUESTED** [review]
Sorry for reviewing it in stages, I haven't started the app until now 🫠

but these bugs could've easily been found with some basic manual testing xd

**Review by Tschonti — APPROVED** [review]
sorry az előző review elég passzív agresszív volt xd

**Pull request merged**